docs/design/coreclr/jit/investigate-stress.md

```diff
@@ -0,0 +1,19 @@
+# Investigating JIT and GC stress
+
+There are two stressing related features for the JIT and JIT generated GC info &ndash; JIT Stress and GC Stress. These features provide a way during development to discover edge cases and more "real world" scenarios without having to develop complex applications.
+
+## JIT Stress (Debug builds only)
+
+Enabling JIT Stress can be done in several ways. Setting `DOTNET_JitStress` to a non-zero integer value that will generate varying levels of JIT optimizations based on a hash of the method's name or set to a value of two (for example, `DOTNET_JitStress=2`) that will apply all optimizations. Another way to enable JIT Stress is by setting `DOTNET_JitStressModeNamesOnly=1` and then requesting the stress modes, space delimited, in the `DOTNET_JitStressModeNames` variable (for example, `DOTNET_JitStressModeNames=STRESS_USE_CMOV STRESS_64RSLT_MUL STRESS_LCL_FLDS`).
+
+It is often useful to use [JIT Dump](./viewing-jit-dumps.md) in tandem with JIT Stress.
+
+## GC Stress
+
+Enabling GC Stress causes GCs to always occur in specific locations. GC Stress can be enabled using the `DOTNET_GCStress` environment variable. It takes a non-zero integer value in hexadecimal format. Note these values can be or'd together (for example, `0x3 = 0x1 | 0x2`).
+
+- **0x1** &ndash; GC on all allocs and 'easy' places.
+- **0x2** &ndash; GC on transitions to Preemptive GC.
+- **0x4** &ndash; GC on every allowable JITed instr.
+- **0x8** &ndash; GC on every allowable R2R instr.
+- **0xF** &ndash; GC only on a unique stack trace.
```

Comment:
Is this true? Isn't 0xF the same as "all of the previous locations"?

You could mention common combos, like 0xC which runs GC stress for each JIT generated instruction (either dynamically or AOT, in R2R). 1 and 2 are "in the VM". Failures in 4 & 8 typically mean a failure in GC info. Failures in 1 & 2 can be due to other VM-related reasons, like lack of GC reporting/pinning in interop frames.

You could mention that for 4 & 8, GC only happens once for any instruction, so can miss failures that only occur on non-first GCs.

You could mention the mechanism for 4 & 8 (replacing with breakpoint instruction), and how that affects disassembly, as well as how stress tries to move the GC heap (but often doesn't succeed, so if there is a hole, it might not get detected even under stress).

You could talk about typical failure modes, such as what underreporting and overreporting are, and typical stack traces. And talk about how to debug GC stress failures, such as using SOS `!u -gcinfo`, JitDump, etc.

Maybe GC stress and JIT stress should get their own separate documents?

Some of these things could be left as TBD.

Reply:
These are all great comments and definitely are educational - especially to me. However, I am learning about from you in real time. I will add some notes about the common combinations but leave it to people more familiar with the area to enrich the document.

> Maybe GC stress and JIT stress should get their own separate documents?

That is something to consider for the future as content increases. Right now I think a single document is sufficient.